common/functions/src/scalars/udfs/in_basic.rs

```diff
@@ -0,0 +1,258 @@
+                }
+                _ => {}
+            }
+        }
+        for idx in 0..$INPUT_ARRAY.len() {
+            let val = $INPUT_ARRAY.try_get(idx)?;
+            match val {
+                DataValue::$CAST_TYPE(Some(val)) => {
+                    let v = match vals_set.contains(&val) {
+                        true => !$NEGATED,
+                        false => $NEGATED,
+                    };
+                    $BUILDER.push(v);
+                }
+                DataValue::$CAST_TYPE(None) => {
+                    $BUILDER.push(false);
+                }
+                _ => {
+                    return Err(ErrorCode::LogicalError("it's a bug"));
+                }
+            }
+        }
+    };
+}
+
+// float type can not impl Hash and Eq trait, so it can not use HashSet
+// maybe we can find some more efficient way to make it.
+macro_rules! float_contains {
+    ($INPUT_DT: expr, $INPUT_ARRAY: expr, $CHECK_ARRAY: expr, $NEGATED: expr, $BUILDER: expr, $CAST_TYPE: ident) => {
+        let mut vals_set = Vec::new();
+        for array in $CHECK_ARRAY {
+            let array = array.column().cast_with_type($INPUT_DT)?;
+            let val = array.try_get(0)?;
+            match val {
+                DataValue::$CAST_TYPE(Some(val)) => {
+                    vals_set.push(val);
+                }
+                _ => {}
+            }
+        }
+        for idx in 0..$INPUT_ARRAY.len() {
+            let val = $INPUT_ARRAY.try_get(idx)?;
+            match val {
+                DataValue::$CAST_TYPE(Some(val)) => {
+                    let v = match vals_set.contains(&val) {
+                        true => !$NEGATED,
+                        false => $NEGATED,
+                    };
+                    $BUILDER.push(v);
+                }
+                DataValue::$CAST_TYPE(None) => {
+                    $BUILDER.push(false);
+                }
+                _ => {
+                    return Err(ErrorCode::LogicalError("it's a bug"));
+                }
+            }
+        }
+    };
+}
+
+impl<const NEGATED: bool> Function for InFunction<NEGATED> {
+    fn name(&self) -> &str {
+        "InFunction"
+    }
+
+    fn return_type(&self, args: &[DataTypeAndNullable]) -> Result<DataTypeAndNullable> {
+        let input_dt = args[0].data_type();
+        if input_dt == &DataType::Null {
+            return Ok(DataTypeAndNullable::create(input_dt, false));
+        }
+
+        let dt = DataType::Boolean;
+        Ok(DataTypeAndNullable::create(&dt, false))
+    }
+
+    fn eval(&self, columns: &DataColumnsWithField, _input_rows: usize) -> Result<DataColumn> {
+        let input_column = columns[0].column();
+
+        let input_array = match input_column {
+            DataColumn::Array(array) => array.to_owned(),
+            DataColumn::Constant(scalar, _) => scalar.to_array()?,
+        };
+
+        let input_dt = input_array.data_type();
+        if input_dt == &DataType::Null {
```

Comment:
I think we cannot cast with the `input_dt` directly. 
Consider the following example: `select 1 in (1.2, 2 )`, If `input_dt` is `DataType::UInt8`,  `1.2` will be cast to `1`, then the result will be `true`.

Reply:
thanks for reminding me, it should not do cast directly here.